Catch disallowed characters in map/mod name

When trying to search for contains "Astro" on the map vault you get an exception, probably due to the client trying to detect a map on the file system, that contains a "?", which is forbidden as a windows file path. We should catch all forbidden windows characters and just not display that map.

https://api.faforever.com/data/map?filter=id==4859

```
java.util.concurrent.CompletionException: java.nio.file.InvalidPathException: Illegal char <?> at index 17: astro_crater_batl?_5vs5.v0002.png
	at java.util.concurrent.CompletableFuture.encodeThrowable(Unknown Source)
	at java.util.concurrent.CompletableFuture.completeThrowable(Unknown Source)
	at java.util.concurrent.CompletableFuture.uniAccept(Unknown Source)
	at java.util.concurrent.CompletableFuture$UniAccept.tryFire(Unknown Source)
	at java.util.concurrent.CompletableFuture.postComplete(Unknown Source)
	at java.util.concurrent.CompletableFuture$AsyncSupply.run(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.lang.Thread.run(Unknown Source)
Caused by: java.nio.file.InvalidPathException: Illegal char <?> at index 17: astro_crater_batl?_5vs5.v0002.png
	at sun.nio.fs.WindowsPathParser.normalize(Unknown Source)
	at sun.nio.fs.WindowsPathParser.parse(Unknown Source)
	at sun.nio.fs.WindowsPathParser.parse(Unknown Source)
	at sun.nio.fs.WindowsPath.parse(Unknown Source)
	at sun.nio.fs.WindowsFileSystem.getPath(Unknown Source)
	at sun.nio.fs.AbstractPath.resolve(Unknown Source)
	at com.faforever.client.remote.AssetServiceImpl.loadAndCacheImage(AssetServiceImpl.java:51)
	at com.faforever.client.remote.AssetServiceImpl.loadAndCacheImage(AssetServiceImpl.java:37)
	at com.faforever.client.map.MapServiceImpl.loadPreview(MapServiceImpl.java:331)
	at com.faforever.client.map.MapServiceImpl.loadPreview(MapServiceImpl.java:327)
	at sun.reflect.GeneratedMethodAccessor133.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.lang.reflect.Method.invoke(Unknown Source)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:333)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:190)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:157)
	at org.springframework.cache.interceptor.CacheInterceptor$1.invoke(CacheInterceptor.java:52)
	at org.springframework.cache.interceptor.CacheAspectSupport.invokeOperation(CacheAspectSupport.java:345)
	at org.springframework.cache.interceptor.CacheAspectSupport.execute(CacheAspectSupport.java:408)
	at org.springframework.cache.interceptor.CacheAspectSupport.execute(CacheAspectSupport.java:327)
	at org.springframework.cache.interceptor.CacheInterceptor.invoke(CacheInterceptor.java:61)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:213)
	at com.sun.proxy.$Proxy76.loadPreview(Unknown Source)
	at com.faforever.client.map.MapCardController.setMap(MapCardController.java:79)
	at com.faforever.client.map.MapVaultController.lambda$appendSearchResult$16(MapVaultController.java:287)
	at java.util.stream.ReferencePipeline$3$1.accept(Unknown Source)
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(Unknown Source)
	at java.util.stream.AbstractPipeline.copyInto(Unknown Source)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(Unknown Source)
	at java.util.stream.ReduceOps$ReduceTask.doLeaf(Unknown Source)
	at java.util.stream.ReduceOps$ReduceTask.doLeaf(Unknown Source)
	at java.util.stream.AbstractTask.compute(Unknown Source)
	at java.util.concurrent.CountedCompleter.exec(Unknown Source)
	at java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
	at java.util.concurrent.ForkJoinPool$WorkQueue.pollAndExecCC(Unknown Source)
	at java.util.concurrent.ForkJoinPool.helpComplete(Unknown Source)
	at java.util.concurrent.ForkJoinPool.externalHelpComplete(Unknown Source)
	at java.util.concurrent.ForkJoinTask.externalAwaitDone(Unknown Source)
	at java.util.concurrent.ForkJoinTask.doInvoke(Unknown Source)
	at java.util.concurrent.ForkJoinTask.invoke(Unknown Source)
	at java.util.stream.ReduceOps$ReduceOp.evaluateParallel(Unknown Source)
	at java.util.stream.AbstractPipeline.evaluate(Unknown Source)
	at java.util.stream.ReferencePipeline.collect(Unknown Source)
	at com.faforever.client.map.MapVaultController.appendSearchResult(MapVaultController.java:290)
	at com.faforever.client.map.MapVaultController.displayMaps(MapVaultController.java:315)
	... 7 more
```

--- a/src/main/java/com/faforever/client/remote/AssetService.java
+++ b/src/main/java/com/faforever/client/remote/AssetService.java
@@ -13,6 +13,7 @@ import org.springframework.web.util.UriUtils;
 import java.net.URL;
 import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
+import java.nio.file.InvalidPathException;
 import java.nio.file.Path;
 import java.util.function.Supplier;
 
@@ -46,18 +47,26 @@ public class AssetService {
       return defaultSupplier.get();
     }
 
-    String urlString = url.toString();
-    urlString = urlValidator.isValid(urlString) ? urlString : UriUtils.encodePath(urlString, StandardCharsets.UTF_8);
-    String filename = urlString.substring(urlString.lastIndexOf('/') + 1);
-    Path cachePath = preferencesService.getCacheDirectory().resolve(cacheSubFolder).resolve(filename);
-    if (Files.exists(cachePath)) {
-      log.debug("Using cached image: {}", cachePath);
-      return new Image(noCatch(() -> cachePath.toUri().toURL().toExternalForm()), width, height, true, true);
-    }
+    try {
+      String urlString = url.toString();
+      urlString = urlValidator.isValid(urlString) ? urlString : UriUtils.encodePath(urlString, StandardCharsets.UTF_8);
+      String filename = urlString.substring(urlString.lastIndexOf('/') + 1);
+      Path cachePath = preferencesService.getCacheDirectory().resolve(cacheSubFolder).resolve(filename);
+      if (Files.exists(cachePath)) {
+        log.debug("Using cached image: {}", cachePath);
+        return new Image(noCatch(() -> cachePath.toUri().toURL().toExternalForm()), width, height, true, true);
+      }
 
-    log.debug("Fetching image {}", url);
-    Image image = new Image(urlString, width, height, true, true, true);
-    JavaFxUtil.persistImage(image, cachePath, filename.substring(filename.lastIndexOf('.') + 1));
-    return image;
+      log.debug("Fetching image {}", url);
+      Image image = new Image(urlString, width, height, true, true, true);
+      JavaFxUtil.persistImage(image, cachePath, filename.substring(filename.lastIndexOf('.') + 1));
+      return image;
+    } catch (InvalidPathException e) {
+      log.warn("Unable to load image due to invalid fileName {}", url, e);
+      if (defaultSupplier == null) {
+        return null;
+      }
+      return defaultSupplier.get();
+    }
   }
 }

